Reimplement grouping as analysis input

MBartkowiakSTFC opened the pull request:
**Description of work**
Following user feedback, new grouping mechanism has been introduced. The main intention was to make it compatible with the new atom selection mechanism.

Affected analysis types:
Elastic Incoherent Structure Factor
Dynamic Incoherent Structure Factor
Gaussian Dynamic Incoherent Structure Factor
Position Autocorrelation Function
Mean Square Displacement
Root Mean Square Deviation

More work will be needed in the future to enable grouping in the Root Mean Square Fluctuation analysis.

closes #644
closes #680

**Fixes**
1. Apart from the default "atom", two new grouping options are "average over molecules" and "individual molecules".
2. GroupingTool class has been introduced to assign the correct results to their corresponding output datasets.
3. The atom-averaged results are always included. The molecule-grouped results are produced from the same calculations, in order to reduce repetition in the calculations.
4. The analysis jobs still iterate over individual atom indices, same as for the "atom" grouping.
5. Unit tests have been added for the analysis types that use grouping.
6. At the moment, molecules which are not fully within the selection are **omitted** from the calculation.

**To test**
Load a trajectory that contains molecules. Generate a Centre of Mass Trajectory out of it.
Run the jobs listed above on the COM trajectory (using "atom" grouping"), and on the normal trajectory using the molecule grouping. The results of the COM analysis and molecule-grouped analysis should be close to each other, but not the same.

MBartkowiakSTFC committed:
Create a grouping tool

MBartkowiakSTFC committed:
Switch MeanSquareDisplacement to GroupingTool

MBartkowiakSTFC committed:
Add docstrings to GroupingTool

MBartkowiakSTFC committed:
Switch EISF PACF and RMSD to GroupingTool

MBartkowiakSTFC committed:
Switch DISF to GroupingTool

MBartkowiakSTFC committed:
Switch GDISF to GroupingTool

MBartkowiakSTFC committed:
Enable molecule grouping runs

MBartkowiakSTFC committed:
Add centre-of-mass averaging to GroupingTool

MBartkowiakSTFC committed:
Improve grouped results for DISF and GDISF

MBartkowiakSTFC committed:
Add unit tests for grouping

MBartkowiakSTFC committed:
Warn on incompletely selected molecules

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py:
```diff
@@ -66,6 +67,10 @@ def initialise_atoms(self, element_list: List[str], name_list: List[str] = None)
         if name_list is not None:
             self._atom_names = [str(x) for x in name_list]
+        for element in self._unique_elements:
+            self.element_indices[element] = {
```
Bear in mind that `set` is unordered. Doubt this will cause problems here, though.

MBartkowiakSTFC replied:
You're right. Here we only use the sets to check which index belongs for which set, so it should be fine.

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+
+        """
+        self._weight_dictionary = {
+            "".join([*key]): value for key, value in new_weights.items()
```
```suggestion
            "".join(key): value for key, value in new_weights.items()
```
If key is a sequence this will do.

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+        """
+        self._grouping = GROUPING_LABELS.get(text_key, GroupingLevel.ATOM)
+
+    def set_weight_dictionary(self, new_weights: dict[tuple[str], float]):
```
```suggestion
    def set_weight_dictionary(self, new_weights: dict[tuple[str, ...], float]):
```
If it is an arbitrary sized tuple
```suggestion
    def set_weight_dictionary(self, new_weights: dict[tuple[str, str], float]):
```
If it is fixed size (in this case 2)

MBartkowiakSTFC replied:
The way MDANSE works, in all the jobs that use the GroupingTool this will be a fixed-size tuple containing a single string.

oerc0122 replied:
In which case this is fine, just wanted to double check because that's an odd use of a tupe.

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+        """
+        for key in self._mandatory_keys:
+            if key not in parameters:
+                raise KeyError(f"Setting {key} is missing from data parameters.")
```
```suggestion
        if missing := self._mandatory_keys - parameters.keys():
            raise KeyError(f"Settings ({', '.join(missing)}) are missing from data parameters.")
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+            value = parameters.get(key)
+            if value is None:
+                value = BACKUP_DATA_PARAMETERS[key]
+            self._data_parameters[key] = value
```
```suggestion
            self._data_parameters[key] = parameters.get(key, BACKUP_DATA_PARAMETERS[key])
```

oerc0122 replied:
In fact there's no need to have two loops here. You've guaranteed the mandatory keys are in the keys, so we can instead just make this:

```suggestion
        for key in self._mandatory_keys + self._extra_keys:
            self._data_parameters[key] = parameters.get(key, BACKUP_DATA_PARAMETERS.get(key))
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+    "average over molecules": GroupingLevel.MOL_AVERAGE,
+}
+
+BACKUP_DATA_PARAMETERS = {
```
I would call these `DEFAULT` rather than `BACKUP`

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+        pardict = {key: self._data_parameters[key] for key in self._extra_keys}
+        if override:
+            for key, value in override.items():
+                pardict[key] = value
```
```suggestion
            pardict.update(override)
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+        """
+        self._debug_counter[name] += 1
+        pardict = {key: self._data_parameters[key] for key in self._extra_keys}
+        if override:
```
May want to check:
```suggestion
        if override is not None:
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+        if self._grouping == GroupingLevel.MOL_AVERAGE:
+            self.create_averaged_molecule_groups(name)
+        elif self._grouping == GroupingLevel.MOL_EACH:
+            self.create_individual_molecule_groups(name)
```
```suggestion
        if self._grouping is GroupingLevel.MOL_AVERAGE:
            self.create_averaged_molecule_groups(name)
        elif self._grouping is GroupingLevel.MOL_EACH:
            self.create_individual_molecule_groups(name)
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+                self._cs.element_indices[atom_type]
+            )
+            if indices:
+                dset_name = "_".join([name, str(atom_type)])
```
```suggestion
                dset_name = f"{name}_{atom_type}"
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+                self._output_data[dset_name].atom_indices = list(indices)
+                self._indices_per_data_key[dset_name] = indices
+                self._plain_datasets.add(dset_name)
+        dset_name = "_".join([name, "total"])
```
```suggestion
        dset_name = f"{name}_total"
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+                elif len(trimmed_mol):
+                    molecules_dropped.append(mindex)
+            if all_indices:
+                dset_name = "_".join([name, str(molecule), "all"])
```
```suggestion
                dset_name = f"{name}_{molecule}_all"
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+            for mindex, mol in enumerate(self._cs._clusters[molecule]):
+                trimmed_mol = self._current_selection.intersection(mol)
+                if set(mol) == trimmed_mol:
+                    all_indices.update(mol)
```
```suggestion
                    all_indices |= mol
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+            for mindex, mol in enumerate(self._cs._clusters[molecule]):
+                trimmed_mol = self._current_selection.intersection(mol)
+                if set(mol) == trimmed_mol:
+                    dset_name = "_".join([name, str(molecule), str(mindex + 1)])
```
```suggestion
                    dset_name = f"{name}_{molecule}_{mindex + 1}"
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+                trimmed_mol = self._current_selection.intersection(mol)
+                if set(mol) == trimmed_mol:
+                    all_indices.update(mol)
+                elif len(trimmed_mol):
```
```suggestion
                elif trimmed_mol:
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+                    self._output_data[dset_name].atom_indices = list(trimmed_mol)
+                    self._indices_per_data_key[dset_name] = trimmed_mol
+                    self._molecule_datasets.add(dset_name)
+                elif len(trimmed_mol):
```
```suggestion
                elif trimmed_mol:
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+        self,
+        index: int,
+        result: np.ndarray,
+        normalise: bool = True,
```
Generally considered good practise to make `bool`s kw-only.

```suggestion
        *,
        normalise: bool = True,
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+            self.assign_result_atomic(index, result, normalise)
+
+    def assign_result_molecular(
+        self, index: int, result: np.ndarray, centre_of_mass: bool = True
```
```suggestion
        self, index: int, result: np.ndarray, *, centre_of_mass: bool = True
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+                    self._molecule_mass[dset_name] += 1.0
+
+    def assign_result_atomic(
+        self, index: int, result: np.ndarray, normalise: bool = True
```
```suggestion
        self, index: int, result: np.ndarray, *, normalise: bool = True
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Formats/HDFFormat.py:
```diff
@@ -113,14 +113,18 @@ def write(
             dset = outputFile.create_dataset(varName, data=var, shape=var.shape)
 
             # All the attributes stored in the OutputVariable instance are written to the HDF file.
             for k, v in list(vars(var).items()):
```
```suggestion
            for k, v in vars(var).items():
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Formats/HDFFormat.py:
```diff
@@ -113,14 +113,18 @@ def write(
+                    dset.attrs[k] = v
+                except TypeError:
+                    print(f"type error: {k}: {v} in {key}")
+                    raise RuntimeError("Bad HDF5 write")
```
```suggestion
                except TypeError as error:
                    log.ERROR("TypeError: %s: %s in %s", k, v, key)
                    raise RuntimeError("Bad HDF5 write.") from error
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/DynamicIncoherentStructureFactor.py:
```diff
@@ -113,11 +108,67 @@ def initialize(self):
+        self._groupers["sqf"] = GroupingTool(
+            self.configuration["trajectory"]["instance"].chemical_system,
+            self._outputData,
+        )
```
```suggestion
        self._groupers = {
            "fqt": GroupingTool(
                self.configuration["trajectory"]["instance"].chemical_system,
                self._outputData,
            ),
            "sqf": GroupingTool(
                self.configuration["trajectory"]["instance"].chemical_system,
                self._outputData,
            ),
        }
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/DynamicIncoherentStructureFactor.py:
```diff
@@ -259,10 +260,10 @@ def combine(self, index, disf_per_q_shell):
-            self._outputData[f"f(q,t)_{element}"][i, :] += v
+        self._groupers["fqt"].assign_result(
+            self.configuration["atom_selection"]["flatten_indices"][index],
+            np.array([v for v in disf_per_q_shell.values()]),
```
```suggestion
            np.array(list(disf_per_q_shell.values())),
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/DynamicIncoherentStructureFactor.py:
```diff
@@ -271,50 +272,29 @@ def finalize(self):
+        self._groupers["fqt"].finalise_centre_of_mass()
+
+        for inkey in self._outputData.keys():
+            if "f(q,t)" in inkey:
```
```suggestion
        for inkey in self._outputData:
            if "f(q,t)" in inkey:
```
Could also write as:
```suggestion 
        for inkey in (key for key in self._outputData if "f(q,t)" in key):
```
And remove a layer of nesting

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/DynamicIncoherentStructureFactor.py:
```diff
@@ -271,50 +272,29 @@ def finalize(self):
+
+        for inkey in self._outputData.keys():
+            if "f(q,t)" in inkey:
+                outkey = "_".join(["s(q,f)"] + inkey.split("_")[1:])
```
```suggestion
                outkey = "s(q,f)_{inkey.split('_', 1)[1]}"
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/DynamicIncoherentStructureFactor.py:
```diff
@@ -271,50 +272,29 @@ def finalize(self):
-            )
-
+                if self.add_ideal_results:
+                    outkey = "_".join(["s(q,f)_ideal"] + inkey.split("_")[1:])
```
```suggestion
                    outkey = "s(q,f)_ideal_{inkey.split('_', 1)[1]}"
```
Although I'd probably recommend making `inkey.split("_", 1)[1]` a named variable since you're using it twice

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/DynamicIncoherentStructureFactor.py:
```diff
@@ -271,50 +272,29 @@ def finalize(self):
+                        axis=1,
+                    )
+
+        for group in self._groupers:
+            if group != "fqt":
```
Can remove layer of nesting
```suggestion
        for group in filter(lambda x: x != "fqt", self._groupers):
```
Or perhaps more pythonic, this also returns the grouper directly meaning we don't need to `getitem` from `self._groupers`
```suggestion
        for grouper in (value for key, value in self._groupers.items() if key != "fqt"):
```

oerc0122 replied:
Can remove layer of nesting
```suggestion
        for group in filter(lambda x: x != "fqt", self._groupers):
```
Or perhaps more pythonic, this also returns the grouper directly meaning we don't need to `getitem` from `self._groupers`
```suggestion
        for grouper in (value for key, value in self._groupers.items() if key != "fqt"):
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/ElasticIncoherentStructureFactor.py:
```diff
@@ -97,10 +92,26 @@ def initialize(self):
         Initialize the input parameters and analysis self variables
         """
         super().initialize()
+        self._nQShells = self.configuration["q_vectors"]["n_shells"]
 
         self.numberOfSteps = self.configuration["atom_selection"]["selection_length"]
```
Any reason for re-ordering this?

MBartkowiakSTFC replied:
The only reason was that it had to be set before the call to `self._grouper.set_dataset_parameters`. Any other position in the code before that line would have worked too.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/GaussianDynamicIncoherentStructureFactor.py:
```diff
@@ -150,69 +145,63 @@ def initialize(self):
+            {atom: 1.0 for atom in nAtomsPerElement}, nAtomsPerElement, 1
+        )
+
+        self._groupers = {}
+        for key in ["fqt", "sqf", "msd"]:
+            self._groupers[key] = GroupingTool(
+                self.configuration["trajectory"]["instance"].chemical_system,
+                self._outputData,
             )
```
```suggestion
        self._groupers = {
            key: GroupingTool(
                self.configuration["trajectory"]["instance"].chemical_system,
                self._outputData,
            ) 
            for key in ("fqt", "sqf", "msd")
        }
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/GaussianDynamicIncoherentStructureFactor.py:
```diff
@@ -267,67 +256,43 @@ def combine(self, index: int, x: tuple[np.ndarray, np.ndarray]):
+
+        for inkey in self._outputData.keys():
+            if "f(q,t)" in inkey:
+                outkey = "_".join(["s(q,f)"] + inkey.split("_")[1:])
```
See above

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/PositionAutoCorrelationFunction.py:
```diff
@@ -98,22 +109,21 @@ def initialize(self):
+        weights = self.configuration["weights"].get_weights()
+        weight_dict = get_weights(weights, nAtomsPerElement, 1)
+        self._grouper.set_weight_dictionary(weight_dict)
+        print(weight_dict)
```
Leftover print?

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/RootMeanSquareDeviation.py:
```diff
@@ -87,29 +100,16 @@ def initialize(self):
             "instance"
         ].chemical_system.atom_list
+        nAtomsPerElement = self.configuration["atom_selection"].get_natoms()
+        weight_dict = {key: 1.0 for key in nAtomsPerElement}
```
```suggestion
        weight_dict = dict.fromkeys(nAtomsPerElement, 1.0)
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/GroupingTool.py:
```diff
@@ -0,0 +1,415 @@
+                norm = 1.0
+            self._output_data[dset_name] /= norm
+            if not np.isclose(exponent, 1.0):
+                self._output_data[dset_name][:] = (
+                    self._output_data[dset_name] ** exponent
+                )
```
It should be noted that there may be optimisations in place for e.g. `sqrt`/`cbrt`, which cannot be done for generalised exponents.

MBartkowiakSTFC replied:
Since 0.5 is what we really expect here, I added a check for it. For exponent 0.5 we will use `np.sqrt`.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/OutputVariables/IOutputVariable.py:
```diff
@@ -61,6 +61,7 @@ def __new__(
         *,
         main_result: bool = False,
         partial_result: bool = False,
+        dtype: type = np.float64,
```
Strictly these should be of type [`np.dtype`](https://numpy.org/doc/2.2/reference/arrays.dtypes.html)

Not sure it matters too much.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/OutputVariables/IOutputVariable.py:
```diff
@@ -87,9 +88,9 @@ def __new__(
         """
         if isinstance(value, tuple):
-            value = np.zeros(value, dtype=np.float64)
+            value = np.zeros(value, dtype=dtype)
         else:
-            value = np.array(list(value), dtype=np.float64)
+            value = np.array(list(value), dtype=dtype)
```
This will not correctly convert nested data.

MBartkowiakSTFC replied:
Please tell me more about what scenario will make it fail.

oerc0122 replied:
If you had a generator of generators, this will only convert the top level. Not sure if this is a likely case coming in.

oerc0122 reviewed (commented):
A few things to consider.

MBartkowiakSTFC committed:
Apply suggestions from code review

Co-authored-by: Jacob Wilkins <46597752+oerc0122@users.noreply.github.com>
Signed-off-by: Maciej Bartkowiak <108934199+MBartkowiakSTFC@users.noreply.github.com>

MBartkowiakSTFC committed:
Merge branch 'protos' of https://github.com/ISISNeutronMuon/MDANSE into maciej/atom-grouping

MBartkowiakSTFC committed:
Merge branch 'maciej/atom-grouping' of https://github.com/ISISNeutronMuon/MDANSE into maciej/atom-grouping

MBartkowiakSTFC committed:
Implement suggestions from code review

MBartkowiakSTFC committed:
Correct set operations in GroupingTool

MBartkowiakSTFC committed:
Correct the log entry in HDFFormat

ChiCheng45 commented:
I've run a GDISF calculation using the DFTB water trajectory. I get a different result for `f(q,t)_total` and `f(q,t)_H2_O1_all`. 

<img width="616" alt="image" src="https://github.com/user-attachments/assets/743d091a-7798-414c-95f9-1b8e468c93c9" />


I know that the grouping level is `average over molecules` but I don't think it make sense for grouping to work like that for this and other similar analysis jobs. The total intermediate scattering function is

```math
F_{\mathrm{inc}, \mathrm{tot}}(q, t) = \frac{1}{N} \sum_{j} w_j \langle \exp[-iq \cdot r_{j}(0)]\exp[iq \cdot r_{j}(t)] \rangle 
```

when we use the atom grouping, the above sum is split up by atom.

```math
F_{\mathrm{inc}, \mathrm{tot}}(q, t) = \sum_{\alpha} W_{\alpha} F_{\mathrm{inc}, \alpha}(q, t) 
```

```math
F_{\mathrm{inc}, \alpha}(q, t) = \frac{1}{Nc_{\alpha}} \sum_{j \in \alpha} \langle \exp[-iq \cdot r_{j}(0)]\exp[iq \cdot r_{j}(t)] \rangle 
```

For the molecular case, we could split the summation into atomic parts associated to each molecule. For a water and ethanol system, we would get 

```math
F_{\mathrm{inc}, \mathrm{tot}}(q, t) = F_{\mathrm{inc}, \mathrm{H2O}}(q, t) + F_{\mathrm{inc}, \mathrm{EtOH}}(q, t) 
```

where

```math
F_{\mathrm{inc}, \mathrm{H2O}}(q, t) = W^{\mathrm{H2O}}_{\mathrm{H}}F_{\mathrm{inc}, \mathrm{H}}^{\mathrm{H2O}}(q, t) + W^{\mathrm{H2O}}_{\mathrm{O}}F_{\mathrm{inc}, \mathrm{O}}^{\mathrm{H2O}}(q, t)
```

and

```math
F_{\mathrm{inc}, \mathrm{EtOH}}(q, t) = W^{\mathrm{EtOH}}_{\mathrm{H}}F_{\mathrm{inc}, \mathrm{H}}^{\mathrm{EtOH}}(q, t) + W^{\mathrm{EtOH}}_{\mathrm{C}}F_{\mathrm{inc}, \mathrm{C}}^{\mathrm{EtOH}}(q, t) + W^{\mathrm{EtOH}}_{\mathrm{O}}F_{\mathrm{inc}, \mathrm{O}}^{\mathrm{EtOH}}(q, t)
```

The intermediate scattering functions are

```math
F_{\mathrm{inc}, \mathrm{H}}^{\mathrm{H2O}}(q, t) = \frac{1}{Nc^{\mathrm{H2O}}_{\mathrm{H}}} \sum_{j \in (\mathrm{H2O} \cap \mathrm{H})} \langle \exp[-iq \cdot r_{j}(0)]\exp[iq \cdot r_{j}(t)] \rangle
```
and 

```math
W^{\mathrm{H2O}}_{\mathrm{H}} = c_{\mathrm{H}}^{\mathrm{H2O}} w_{\mathrm{H}}
```

where $c_{\mathrm{H}}^{\mathrm{H2O}} = N_{\mathrm{H}}^{\mathrm{H2O}} / N$ and $N_{\mathrm{H}}^{\mathrm{H2O}}$ is the number of hydrogen atoms in water. 


For the water trajectory above my expectations would've been 

```math
F_{\mathrm{inc}, \mathrm{tot}}(q, t) = F_{\mathrm{inc}, \mathrm{H2O}}(q, t)
```

so that the plots above should be the same since I had the `Apply weights?` checked.

MBartkowiakSTFC commented:
Thanks for this comment. If we implemented grouping the way you described it would certainly make more sense to me too.

In the code of MDANSE prior to this PR, any grouping of the results other than "atom" would result in MDANSE taking several atoms, then calculating the trajectory of _their centre of mass_, and running the analysis on this. This, of course, makes me wonder why we needed grouping to start with, since it is possible to run CenterOfMassTrajectory first and then run the further analysis on it.

@sanghamitra-mukhopadhyay What is grouping supposed to do?
1. Split the total signal into proportional contributions from specific molecules or specific molecule types?
2. Group atoms into averaged objects with averaged dynamics, which can then be analysed?
3. Something else?
You are welcome to give us specific description for the two separate cases, grouping into individual molecules and averaging over molecules of each type.

MBartkowiakSTFC commented:
This has been replaced by #810
